### DISCUSSION TITLE
Grid seems like not rendering grid properly

### DISCUSSION THREAD

**Discussion by WingsDevelopment** [tailwindlabs/tailwindcss#9754] [Help]
I am trying to create GridContainer and GridItem components to implement tailwind grid inside of them, but it seems like I am getting very strange behaviour.. Commented code is what I am trying to accomplish inside my components. And it works on first load, but when i play with my props for example, if i change first two items from 6, 6 to 3, 9 and save it may not change on screen, or it may render it with wrong sizes.. But when i uncomment code above and save, then it reacts on the change and updates the view both displaying new div(from comment) but also updates the changes from the code below... Any idea what's going on ?
version: "tailwindcss": "^3.2.1",

 Here is my code:

```
import React, { PropsWithChildren } from "react";

export const Test = () => {
  return (
    <div className="w-full">
      {/* <div className="grid grid-cols-12 gap-1">
            <div className="col-span-6 border">test</div>
            <div className="col-span-6 border">test</div>
            <div className="col-span-6 border">test</div>
            <div className="col-span-6 border">test</div>
          </div> */}
      <GridContainer>
        <GridItem size={6}>
          <div className="border">test</div>
        </GridItem>
        <GridItem size={6}>
          <div className="border">test</div>
        </GridItem>
        <GridItem size={6}>
          <div className="border">test</div>
        </GridItem>
        <GridItem size={6}>
          <div className="border">test</div>
        </GridItem>
      </GridContainer>
    </div>
  );
};

const GridContainer: React.FC<PropsWithChildren> = ({ children }) => {
  return <div className="grid grid-cols-12">{children}</div>;
};

interface TestProps extends PropsWithChildren<{}> {
  size?: number;
}
const GridItem: React.FC<TestProps> = ({ size, children }) => {
  return <div className={`border col-span-${size ? size : 1}`}>{children}</div>;
};

export default Test;
```

Here is my tailwind config:

```
function withOpacity(variableName) {
  return ({ opacityValue }) => {
    if (opacityValue !== undefined) {
      return `rgba(var(${variableName}), ${opacityValue})`;
    }
    return `rgb(var(${variableName}))`;
  };
}

/** @type {import('tailwindcss').Config} */
module.exports = {
  mode: "jit",
  darkMode: "class",
  content: [
    "./src/**/*.{js,jsx,ts,tsx}",
    "./node_modules/react-tailwindcss-select/dist/index.esm.js",
  ],

  theme: {
    extend: {
      textColor: {
        skin: {
          base: withOpacity("--color-text-base"),
          muted: withOpacity("--color-text-muted"),
          inverted: withOpacity("--color-text-inverted"),
        },
      },
      backgroundColor: {
        skin: {
          fill: withOpacity("--color-fill"),
          accent: withOpacity("--color-accent"),
          "accent-hover": withOpacity("--color-accent-hover"),
          muted: withOpacity("--color-muted"),
          primary: "var(--color-primary)",
        },
      },
      gradientColorStops: {
        skin: {
          hue: withOpacity("--color-fill"),
        },
      },
    },
  },
  plugins: [],
};
```

**Comment by wongjn** [comment]
Tailwind needs to see the class names as unbroken, full strings, not strings constructed via interpolation or concatenation as you have done here. This is because Tailwind does not run your code and does not know the possibilities of what the code might render - it only sees the files as literal strings (think like if the file ended in `.txt` instead of `.jsx`. Read more about this in [the documentation](https://tailwindcss.com/docs/content-configuration#dynamic-class-names).

**Reply by wongjn** [reply]
> Okay I see, thanks. But how do I overcome this problem then? Do i just write native CSS then ?[^1]

There are several solutions:

## Pass the class name in the component prop

```html
<GridItem size="col-span-6">
  <div className="border">test</div>
</GridItem>
```
```tsx
const GridItem: React.FC<TestProps> = ({ size = 'col-span-1', children }) => {
  return <div className={`border ${size}`}>{children}</div>;
};
```

## Use an inline `style` attribute property

```tsx
const GridItem: React.FC<TestProps> = ({ size = 1, children }) => {
  return <div className="border" style={{ gridColumn: `span ${size} / span ${size}` }}>{children}</div>;
};
```

## Use a CSS variable

```tsx
const GridItem: React.FC<TestProps> = ({ size = 1, children }) => {
  return <div className="border col-span-[var(--size)]" style={{ '--size': size }}>{children}</div>;
};
```

## Safelist the classes that might be used

```js
module.exports = {
  // …
  safelist: [
    'col-span-1',
    'col-span-2',
    // …
    'col-span-11',
    'col-span-12',
  ],
  // …
};
```
OR
```js
module.exports = {
  // …
  safelist: Array(12).fill().map((_, i) => `col-span-${i + 1}`),
  // …
};
```

[^1]: https://github.com/tailwindlabs/tailwindcss/discussions/9754#discussioncomment-4067403

**Reply by BuchananDevOps** [reply]
Yea, I'm not looking to be the bearer of bad news but Tailwind CSS isn't compatible with Next 13 in terms of Grid/Flex. While the methods above could be hacked into server side rendering (SSR), with their new approach to sever components and 'use client' for any objects outside the app directory. 

I spent the better part of the last 16 hours rooting for answers, you're (Tailwind) is going to need a plugin/compiler for server components.

**Reply by wongjn** [reply]
One should be able to add server and client source files to the code paths that Tailwind scans via the `content` file globs to compile all the classes required. If you have a particular use case in mind where this is not working for you, please consider providing a remotely-hosted git repository and I can take a look for you.

**Reply by BuchananDevOps** [reply]
I created glob patterns for all my dependencies already trying to debug the issue. 

Were going to give Next 13.0.5-canary.4 a shot and see if that's the issue.

We're testing for a large deployment/upgrade from React 17 to 18 and Next 12/13. 

There definitely a issue with Grid/Flex on our end when switching getStaticProps to Fetch (Props vs. Params). 

Later this morning I'll install everything fresh and see if if was Next 13.0.4. If it persists I'll set up a public repo for you to reference. 

Oh, thanks for rapid support here.

**Comment by WingsDevelopment** [comment]
Okay I see, thanks.
But how do I overcome this problem then? Do i just write native CSS then ?

**Reply by wongjn** [reply]
Please find my response in the correct thread here: https://github.com/tailwindlabs/tailwindcss/discussions/9754#discussioncomment-4067621
